Anc/adding signup and auth

crowandrew opened the pull request:
### :scroll: Description

- Describe the general shape of this PR (new feature? refactor? bug fix? one-line change?)
This adds two authentication signup and login types for email and password and Google SSO to the GraphQL Backend. It also, adds the framework for a MS SSO. This has multiple changes across most files in the GraphQL backend

- Describe what changes are being made
Added mutation create user accounts with email and password and Google SSO and return a JWT
Added queries to authentication for Local(email and password login) and Google SSO and return a JWT
Locked down all vulnerable routes requiring you to be an ADMIN or USER role.
Added new fields to the account table in PostGres , dropped name, added first name, added last name, added auth type, added password, added password reset token, added, password reset expiration added updated timestamp.
Added framework for Microsoft SSO have some todos that need to be finished.

- Describe why these changes are being made
To allow users and admins to create accounts and to add authentication / authorization to the GraphQL backend.

- List the use cases and edge cases relevant to this PR
New user / admin signup with email and password
New user / admin signup with Google SSO
Login with email and password
Login with Google SSO
Added authorization to vulnerable routes.

- Describe how errors will be handled. How will we know if this code breaks in production
In each query and mutation all errors are logged using logger with specific messaging and an error is returned to the user.

- Describe any external libraries/dependencies added or removed in this PR
bcrypt - for hashing password
jwks-rsa - for verifying MS id tokens

- Describe any security risks are there regarding this change
Currently the upload service is not locked down and the MS signup and login does not verify email addresses. This may also lockout most queries and mutations currently used.

- Describe how you tested these changes
I created multiple local users with both email and password and Google Id Tokens. 
I logged in with each user with either email and password or a Google Id Token
I verified each route was usable with a JWT when needed.

- Link to relevant external documentation



--------------------------
### :clipboard: Mandatory Checklist

- [x] Example of a checked item (please remove when creating your Pull Request) 

- [x] Linked to the Github Issues being addressed using the right sidebar :arrow_right:
- [x] Have you discussed these changes with the project leader(s)?
- [x] Do all variable and function names communicate what they do?
- [x] Were all the changes commented and / or documented?
- [x] Is the PR the right size? (If the PR is too large to review, it might be good to break it up into multiple PRs.)
- [x] Does all work in progress, temporary, or debugger code have a TODO comment with links to Github issues?


--------------------------
### :framed_picture: Screenshots and Screen Recordings

#### Before
<!--- Add before screenshots here -->


#### After
<!--- Add after screenshots here -->


--------------------------
### :blue_book: Glossary
- PR = Pull Request

sent-hil commented on backend/graphql_api/lambda/migrations/0005-account-changes.sql:
```diff
@@ -0,0 +1,24 @@
+drop name;
+
+ALTER TABLE account
+ADD last_name VARCHAR(50);
```
not that I'm asking you to change, but you can do multiple ADD COLUMN & DROP COLUMN under a single ALTER command

crowandrew replied:
I 100% agree. I had written some pseudo-code earlier of what I needed to remove and add. When I wrote the commands I just did them as single commands. Next time I'll refactor it down.

sent-hil commented on backend/graphql_api/lambda/graphql/apolloServer.js:
```diff
@@ -94,13 +100,31 @@ module.exports = (logger) => {
+    }
+  
+
     type auth_data {
```
not sure if a single style is enforced, but shouldn't this be camelCase? i see there's also snack case in type defs and columns so idk

crowandrew replied:
I tried to match the case types in what was already there, and the ALL CAPS for "USER" and "ADMIN was from the examples on here on Enum types. [GrapgQL Schema](https://graphql.org/learn/schema/). I think "Roles" however should be "roles" to match camelCase. Let me know your thoughts. A lot of different case types used I'm fine moving any of this to a single style. I was just trying to match what was there.

crowandrew replied:
Made the changes to roles as states above.

sent-hil commented on backend/graphql_api/lambda/graphql/resolvers/authentication.js:
```diff
@@ -0,0 +1,224 @@
+                throw new Error('Microsoft Id Token is Expired');
+            };
+            //Check if user in database
+            const dbUser = await dbClient.getAccountByEmail(email);
```
noticing some duplication here with https://github.com/codeforsanjose/gov-agenda-notifier/pull/158/files#diff-923e556de69970e65bad376d92d3769de42804d447f2de942f2415ced77ca049R70

crowandrew replied:
Agreed, the Microsoft SSO isn't finished. I need to find a way to verify email and split the name. I will need to refactor this with the Google SSO once I know how much duplication there is in the Microsoft SSO. If you want we can comment out the Microsoft SSO for now. Let me know your thoughts. I wanted to finish the MS SSO before the PR but was running into multiple blockers.

anna-kartynnik commented:
I've just noticed this PR has "64 additions, 23,330 deletions" in `frontend/package-lock.json`. I don't think this PR should touch frontend.

crowandrew commented:
Removed the 

> I've just noticed this PR has "64 additions, 23,330 deletions" in `frontend/package-lock.json`. I don't think this PR should touch frontend.

Just removed `frontend/package-lock.json` from this pull request
